### PR TITLE
MYST-114 / Message signing and verification

### PR DESCRIPTION
**Pull request by ignasbernotas**

### CODE REVIEW

File: identity/keystore_interface.go [hunk 1]
```diff
@@ -5,4 +5,7 @@ import "github.com/ethereum/go-ethereum/accounts"
 type keystoreInterface interface {
 	Accounts() []accounts.Account
 	NewAccount(passphrase string) (accounts.Account, error)
+	Find(a accounts.Account) (accounts.Account, error)
```
Comment (Waldz) [inline identity/keystore_interface.go]: Could You start using `keystore.Find()` in `manager.GetIdentity()` as-well?

File: identity/manager.go [hunk 2]
```diff
@@ -51,6 +54,37 @@ func (idm *identityManager) GetIdentities() []Identity {
 	return ids
 }
 
+func (idm *identityManager) Unlock(identity Identity, passphrase string) error {
+	keystoreManager := idm.keystoreManager
+	accountExisting := identityToAccount(identity)
+
+	account, err := keystoreManager.Find(accountExisting)
```
Comment (Waldz) [inline identity/manager.go]: What if I already have "found" identity which I retrieved with `manager.GetIdentity()`?

File: identity/manager.go [hunk 3]
```diff
@@ -51,6 +54,37 @@ func (idm *identityManager) GetIdentities() []Identity {
...
+	err = keystoreManager.Unlock(account, passphrase)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (idm *identityManager) IsUnlocked(identity Identity) bool {
+	_, err := idm.keystoreManager.SignHash(identityToAccount(identity), messageHash([]byte("1")))
```
Comment (Waldz) [inline identity/manager.go]: - Strange that readonly function does write action
- I would skip implementing `IsUnlocked` flag without real usecase

File: identity/manager.go (hunk 2 above)
Comment (ignasbernotas) [reply identity/manager.go]: What if you didn't? :D

File: identity/signer_test.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	"encoding/hex"
+)
+
+func TestSigning(t *testing.T) {
+	ks := getIdentityTestKeystore()
+	identity := FromAddress("0x1e35193c8cadAA15b43B05ae3D882C91F49BB0Aa")
+	manager := NewIdentityManager(ks)
+	manager.Unlock(identity, "test_passphrase")
+
+	signer := manager.GetSigner(identity)
```
Comment (Waldz) [inline identity/signer_test.go]: You dont need full manager to test only `signer.Sign()`

File: identity/manager.go (hunk 3 above)
Comment (ignasbernotas) [reply identity/manager.go]: I agree, it just seemed like we will need something like this and this will be more friendly than err != keystore.ErrLocked. Plus we won't need to use keystore package externally for these checks:)

File: identity/signer_test.go (hunk 4 above)
Comment (ignasbernotas) [reply identity/signer_test.go]: Accounts still need to be unlocked. How would you approach this?

File: identity/keystore_interface.go (hunk 1 above)
Comment (ignasbernotas) [reply identity/keystore_interface.go]: Yeah, good idea, cheers!

File: identity/manager.go (hunk 2 above)
Comment (Waldz) [reply identity/manager.go]: Seems like we have 2 structures:
- `IdentityAddress` or `IdentityId` - just unique string `0x4e878562376372CCa10397245CBc2950d42246e1`, could point to another node
- `Identity` or `IdentityWallet`- which is my controlled identity from my owned keystore. With Lock/Unlock functions

File: identity/signer.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/ethereum/go-ethereum/crypto"
+	"fmt"
+)
+
+type Signer interface {
+	Sign(message []byte) ([]byte, error)
+}
+
+type keystoreSigner struct {
+	keystoreManager keystoreInterface
```
Comment (Waldz) [inline identity/signer.go]: Just noticed. `Signer` reminds very much structure `keystoreWallet` from `go-ethereum/accounts/accounts`

File: identity/manager.go (hunk 3 above)
Comment (Waldz) [reply identity/manager.go]: And if finally we do have `isLocked` flag, we should read it from Ethereum library see `keystoreWallet.Status()`

File: identity/manager_test.go [hunk 6]
```diff
@@ -3,16 +3,16 @@ package identity
 import (
 	"errors"
 	"testing"
-
 	"github.com/ethereum/go-ethereum/accounts"
 	"github.com/stretchr/testify/assert"
+	"github.com/ethereum/go-ethereum/accounts/keystore"
```
Comment (Waldz) [inline identity/manager_test.go]: This unit was using mocked keystore, we should not need Ethereum library and real keystore in filesystem.

Comment (Waldz) [reply identity/manager_test.go]: If do, would be good good to separate such tests to `integration_test.go`

Comment (ignasbernotas) [reply identity/manager_test.go]: We talked about this with @tadovas. It seemed logical to have some sort of checking whether the underlying ethereum keystore works. What if we update the eth keystore package and something breaks? We won't find out about it with mocks.

File: identity/manager_test.go [hunk 7]
```diff
@@ -89,3 +89,66 @@ func Test_HasIdentity(t *testing.T) {
...
+
+func Test_UnlockError(t *testing.T) {
+	ks := getIdentityTestKeystore()
+	testData := struct {
+		Address    string
+		Passphrase string
+	}{
+		"0x53a835143c0eF3bBCBFa796D7EB738CA7dd28f68",
+		"123",
+	}
```
Comment (donce) [inline identity/manager_test.go]: Is it a common practice to define test data struct even in a single case test?

Comment (ignasbernotas) [reply identity/manager_test.go]: Probably unnecessary:)

File: identity/signer.go (hunk 5 above)
Comment (donce) [inline identity/signer.go]: If it satisfies `keystoreInterface` interface (and not `keystoreManagerInterface`), maybe `Manager` in `keystoreManager` variable name is redundant?

File: identity/verifier.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+func (ev *ethereumVerifier) Verify(message []byte, signature []byte) bool {
+	hash := messageHash(message)
+	pk, err := crypto.Ecrecover(hash, signature)
+	if err != nil {
+		return false
+	}
+	recoveredAddress := crypto.PubkeyToAddress(*crypto.ToECDSAPub(pk)).Hex()
+	return FromAddress(recoveredAddress) == ev.peerIdentity
```
Comment (Waldz) [inline identity/verifier.go]: Just notice. Would be possible to drop attribute `peerIdentity`,
by changing interface:
`Verify(message []byte, signature []byte) bool`
->
`GetSigner(message []byte, signature []byte) Identity`

File: identity/signer.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+		return nil, err
+	}
+
+	return signature, nil
+}
+
+// messageHash is a helper function that calculates a hash for the given message that can be
+// safely used to calculate a signature from.
+//
+// The hash is calulcated as
```
Comment (donce) [inline identity/signer.go]: *calculated*

File: identity/signer.go (hunk 5 above)
Comment (Waldz) [reply identity/signer.go]: Yeah `keystore` is enough

File: identity/signer_test.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+	sig, err := signer.Sign([]byte("Boop!"))
+	assert.NoError(t, err)
+	assert.Equal(
+		t,
+		hex.EncodeToString(sig),
+		"51e8c02f544c20a5b5b92894ffdd4dad90a71d994cad608cb3157b9ed7757de2758b9c0fc51bfaf079a4d60e81cc83c14cf6900c9bc031ba3f44cb119b82a5f000",
+	)
```
Comment (donce) [inline identity/signer_test.go]: Second argument should be `expected`, and the third `actual,` not in reverse (https://godoc.org/github.com/stretchr/testify/assert#Equal):
```go
func Equal(t TestingT, expected, actual interface{}, msgAndArgs ...interface{}) bool
```

File: identity/manager_test.go (hunk 6 above)
Comment (Waldz) [reply identity/manager_test.go]: Yes, agree on such integration test:
- 1 integration with happy path
- 10 units with rare edge cases

File: identity/signer_test.go (hunk 4 above)
Comment (Waldz) [reply identity/signer_test.go]: Usualy tests skip regular factory `New..`.
And loads just dependencies which are tested:
```
keystore := getIdentityTestKeystore()
keystore.Unlock("0x1e35193c8cadAA15b43B05ae3D882C91F49BB0Aa")

signer := keystoreSigner{keystore: keystore}
```

File: identity/verifier_test.go [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+	identity := toIdentity(privateKey)
+	assert.True(t, NewVerifier(identity).Verify(message, signature))
+}
+
+func TestVerifyFunctionReturnsFalseWhenSignatureIsIncorrect(t *testing.T) {
+	message := []byte("I am message!")
+	privateKey, err := crypto.GenerateKey()
+	assert.NoError(t, err)
+
+	signature, err := crypto.Sign(crypto.Keccak256(message), privateKey)
```
Comment (donce) [inline identity/verifier_test.go]: If I'm not mistaken, we could re-use `messageHash` here instead of `crypto.Keccak256`, as in `TestVerifyFunctionReturnsTrueWhenSignatureIsCorrect`. That it would be easier to see the difference between these two tests - only different parts will be different.

File: identity/signer.go [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+func NewSigner(keystore keystoreInterface, identity Identity) Signer {
+	return &keystoreSigner{
+		keystoreManager: keystore,
+		identity:        identity,
+	}
+}
+
+func (ksSigner *keystoreSigner) Sign(message []byte) ([]byte, error) {
+	signature, err := ksSigner.keystoreManager.SignHash(identityToAccount(ksSigner.identity), messageHash(message))
```
Comment (Waldz) [inline identity/signer.go]: If I am signing 100 messages, same `identityToAccount` is called several times for variable `self.identity` is is always same

File: identity/keystore_interface.go (hunk 1 above)
Comment (ignasbernotas) [reply identity/keystore_interface.go]: should fakeKeystore.Find implement the actual find through fake []Accounts? Otherwise the tests fail, but seems redundant to kind of do it.

File: identity/manager.go (hunk 3 above)
Comment (tadovas) [reply identity/manager.go]: If this IsUnlocked functionality already in wallet then we are good to go.
Use case:
1. Consumer gets list of identities from tequila endoint
2. Consumer needs to decide some how if he needs to unlock identity he wants to use for connection. Calling unlock everytime when he wants to use it - sound a bit overhead? lock-status subresource with locked : true| false would be a nice indication in tequila endpoint

File: identity/verifier.go (hunk 8 above)
Comment (tadovas) [reply identity/verifier.go]: I thought about this approach, but then verification from caller side would be a bit clumsy. Also comparing extracted identity with given identity would require a) have a reference to expected identity b) compare it everywhere verification is needed. I think bool return value is cleaner indication that verification succeeded or not

File: identity/verifier_test.go (hunk 11 above)
Comment (tadovas) [reply identity/verifier_test.go]: Missed that. Good point.

File: identity/manager.go [hunk 13]
```diff
@@ -51,6 +54,37 @@ func (idm *identityManager) GetIdentities() []Identity {
...
+	if err != nil {
+		return err
+	}
+
+	err = keystoreManager.Unlock(account, passphrase)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (donce) [inline identity/manager.go]: Instead of these 5 lines of code, we could just return result of `Unlock`:
```go
return keystoreManager.Unlock(account, passphrase)
```
Unless you think it's good to keep `err != nil` check here for consistency?

Comment (ignasbernotas) [reply identity/manager.go]: If we decide to get rid of IsLocked method then maybe we could abstract some of the errors with our own? So that packages using this manager wouldn't have to include ethereum's keystore package just for error constants.

File: identity/manager.go (hunk 3 above)
Comment (ignasbernotas) [reply identity/manager.go]: keystoreWallet is private to the keystore package as far as I can see. Have you seen it exposed through some function?

File: identity/keystore_interface.go (hunk 1 above)
Comment (ignasbernotas) [reply identity/keystore_interface.go]: Fixed.

File: identity/signer_test.go (hunk 10 above)
Comment (ignasbernotas) [reply identity/signer_test.go]: Fixed

File: identity/signer_test.go (hunk 4 above)
Comment (ignasbernotas) [reply identity/signer_test.go]: Fixed

File: identity/signer.go (hunk 5 above)
Comment (ignasbernotas) [reply identity/signer.go]: Yep, but keystoreWallet is private.

Comment (ignasbernotas) [reply identity/signer.go]: Fixed now.

File: identity/signer.go (hunk 9 above)
Comment (ignasbernotas) [reply identity/signer.go]: Fixed.

**Comment by ignasbernotas** [comment]
Test message

File: identity/cache_test.go [hunk 14]
```diff
@@ -18,7 +18,7 @@ func TestIdentityCache_StoreIdentity(t *testing.T) {
 	assert.Nil(t, err)
 }
 
-func Test_IdentityCacheGetIdentity(t *testing.T) {
+func TestIdentityCache_GetIdentity(t *testing.T) {
```
Comment (ignasbernotas) [inline identity/cache_test.go]: test line comment
